detection_rules/rule_loader.py

```diff
@@ -4,132 +4,207 @@
+    """Get a callable that will return true if a dictionary matches a set of criteria.
 
-    if verbose:
-        print("Loading rules from {}".format(RULES_DIR))
+    * each key is a dotted (or __ delimited) path into a dictionary to check
+    * each value is a value or list of values to match
+    """
+    critieria.update(_obj or {})
+    checkers = [(path_getter(k), set(v) if isinstance(v, (list, set, tuple)) else {v}) for k, v in critieria.items()]
 
-    if paths is None:
-        paths = sorted(glob.glob(os.path.join(RULES_DIR, '**', '*.toml'), recursive=True))
+    def callback(obj: dict) -> bool:
+        for getter, expected in checkers:
+            target_values = getter(obj)
+            target_values = set(target_values) if isinstance(target_values, (list, set, tuple)) else {target_values}
 
-    for rule_file in paths:
-        try:
-            # use pytoml instead of toml because of annoying bugs
-            # https://github.com/uiri/toml/issues/152
-            # might also be worth looking at https://github.com/sdispater/tomlkit
-            with io.open(rule_file, "r", encoding="utf-8") as f:
-                file_lookup[rule_file] = pytoml.load(f)
-        except Exception:
-            print(u"Error loading {}".format(rule_file))
-            raise
+            return bool(expected.intersection(target_values))
 
-    if verbose:
-        print("Loaded {} rules".format(len(file_lookup)))
-    return file_lookup
+        return False
 
+    return callback
 
-@cached
-def load_rules(file_lookup=None, verbose=True, error=True):
-    """Load all the rules from toml files."""
-    file_lookup = file_lookup or load_rule_files(verbose=verbose)
-
-    failed = False
-    rules: List[TOMLRule] = []
-    errors = []
-    queries = []
-    query_check_index = []
-    rule_ids = set()
-    rule_names = set()
-
-    for rule_file, rule_contents in file_lookup.items():
+
+def metadata_filter(**metadata) -> Callable[[TOMLRule], bool]:
+    """Get a filter callback based off rule metadata"""
+    flt = dict_filter(metadata)
+
+    def callback(rule: TOMLRule) -> bool:
+        target_dict = rule.contents.metadata.to_dict()
+        return flt(target_dict)
+
+    return callback
+
+
+production_filter = metadata_filter(maturity="production")
+deprecate_filter = metadata_filter(maturity="deprecated")
+
+
+class RuleCollection:
+    """Collection of rule objects."""
+
+    __default = None
+
+    def __init__(self, rules: Optional[List[TOMLRule]] = None):
+        self.id_map: Dict[definitions.UUIDString, TOMLRule] = {}
+        self.file_map: Dict[Path, TOMLRule] = {}
+        self.rules: List[TOMLRule] = []
+        self.frozen = False
+
+        self._toml_cache: Dict[Path, dict] = {}
+
+        for rule in (rules or []):
+            self.add_rule(rule)
+
+    def __len__(self):
+        """Get the total amount of rules in the collection."""
+        return len(self.rules)
+
+    def __iter__(self):
```

Comment:
i thought these were a nice touch and meant more interoperability with existing code.